File: src/main/java/pw/switchcraft/plethora/gameplay/PlethoraFakePlayer.java
package pw.switchcraft.plethora.gameplay;

import com.mojang.authlib.GameProfile;
import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.block.Blocks;
import net.minecraft.block.Material;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityDimensions;
import net.minecraft.entity.EntityPose;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.text.HoverEvent;
import net.minecraft.text.LiteralText;
import net.minecraft.util.Pair;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.world.World;
import pw.switchcraft.plethora.Plethora;
import pw.switchcraft.plethora.api.Constants;
import pw.switchcraft.plethora.mixin.EntityAccessor;
import pw.switchcraft.plethora.mixin.ServerPlayerInteractionManagerAccessor;
import pw.switchcraft.plethora.util.FakePlayer;

import javax.annotation.Nonnull;
import java.lang.ref.WeakReference;

import static net.minecraft.network.packet.c2s.play.PlayerActionC2SPacket.Action.START_DESTROY_BLOCK;

public class PlethoraFakePlayer extends FakePlayer {
    public static final GameProfile PROFILE = new GameProfile(Constants.FAKEPLAYER_UUID, "[" + Plethora.MOD_ID + "]");

    private final WeakReference<Entity> owner;

    private BlockPos digPosition;
    private Block digBlock;

    private int currentDamage = -1;
    private int currentDamageState = -1;

    public PlethoraFakePlayer(ServerWorld world, Entity owner, GameProfile profile) {
        super(world, profile != null && profile.isComplete() ? profile : PROFILE);
        if (owner != null) {
            setCustomName(owner.getName());
            this.owner = new WeakReference<>(owner);
        } else {
            this.owner = null;
        }

        ((EntityAccessor) this).setStandingEyeHeight(0.0f);
    }

    @Nonnull
    @Override
    protected HoverEvent getHoverEvent() {
        Entity owner = getOwner();
        if (owner != null) {
            return new HoverEvent(HoverEvent.Action.SHOW_ENTITY, new HoverEvent.EntityContent(
                owner.getType(),
                owner.getUuid(),
                owner.getName()
            ));
        } else {
            return new HoverEvent(HoverEvent.Action.SHOW_TEXT,
                new LiteralText("PlethoraFakePlayer - No owner!"));
        }
    }

    public Entity getOwner() {
        return owner == null ? null : owner.get();
    }

    @Override
    public float getPitch(float tickDelta) {
        // No pitch lerping
        return this.getPitch();
    }

    @Override
    public float getYaw(float tickDelta) {
        // No yaw lerping, and don't use head yaw like livingEntity does
        return this.getYaw();
    }

    @Override
    public double getEyeY() {
        return this.getY();
    }

    @Override
    public float getEyeHeight(EntityPose pose) {
        return 0.0f;
    }

    @Override
    public float getActiveEyeHeight(EntityPose pose, EntityDimensions dimensions) {
        return 0.0f;
    }

    public void updateCooldown() {
        handSwingTicks = 20;
    }

    private void setState(Block block, BlockPos pos) {
        ServerPlayerInteractionManagerAccessor spim = (ServerPlayerInteractionManagerAccessor) interactionManager;

        spim.setMining(false);
        spim.setBlockBreakingProgress(-1);

        digPosition = pos;
        digBlock = block;
        currentDamage = -1;
        currentDamageState = -1;
    }

    public Pair<Boolean, String> dig(BlockPos pos, Direction side) {
        World world = getEntityWorld();
        BlockState state = world.getBlockState(pos);
        Block block = state.getBlock();
        Material material = state.getMaterial();

        int topY = world.getTopY();

        if (block != digBlock || !pos.equals(digPosition)) setState(block, pos);

        if (!world.isAir(pos) && !material.isLiquid()) {
            if (block == Blocks.BEDROCK || state.getHardness(world, pos) <= -1) {
                return new Pair<>(false, "Unbreakable block detected");
            }

            ServerPlayerInteractionManagerAccessor spim = (ServerPlayerInteractionManagerAccessor) interactionManager;
            for (int i = 0; i < 10; i++) {
                if (currentDamageState == -1) {
                    interactionManager.processBlockBreakingAction(pos, START_DESTROY_BLOCK, side, topY);
                    currentDamageState = spim.getBlockBreakingProgress();
                } else {
                    currentDamage++;
                    float hardness = state.calcBlockBreakingDelta(this, world, pos) * (currentDamage + 1);
                    int hardnessState = (int) (hardness * 10);

                    if (hardnessState != currentDamageState) {
                        world.setBlockBreakingInfo(getId(), pos, hardnessState);
                        currentDamageState = hardnessState;
                    }

                    if (hardness >= 1) {
                        interactionManager.tryBreakBlock(pos);

                        setState(null, null);
                        break;
                    }
                }
            }

            return new Pair<>(true, "block");
        }

        return new Pair<>(false, "Nothing to dig here");
    }
}

File: src/main/java/io/sc3/plethora/gameplay/PlethoraFakePlayer.kt
package io.sc3.plethora.gameplay

import com.mojang.authlib.GameProfile
import io.sc3.plethora.Plethora.modId
import io.sc3.plethora.api.Constants.FAKEPLAYER_UUID
import io.sc3.plethora.mixin.EntityAccessor
import io.sc3.plethora.mixin.ServerPlayerInteractionManagerAccessor
import net.fabricmc.fabric.api.entity.FakePlayer
import net.minecraft.block.Block
import net.minecraft.block.Blocks
import net.minecraft.entity.Entity
import net.minecraft.entity.EntityDimensions
import net.minecraft.entity.EntityPose
import net.minecraft.network.packet.c2s.play.PlayerActionC2SPacket.Action.START_DESTROY_BLOCK
import net.minecraft.server.world.ServerWorld
import net.minecraft.text.HoverEvent
import net.minecraft.text.HoverEvent.Action.SHOW_ENTITY
import net.minecraft.text.HoverEvent.Action.SHOW_TEXT
import net.minecraft.text.Text
import net.minecraft.util.Pair
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import java.lang.ref.WeakReference

class PlethoraFakePlayer(
  world: ServerWorld,
  ownerEntity: Entity?,
  profile: GameProfile?
) : FakePlayer(
  world,
  if (profile != null && profile.isComplete) profile else PROFILE
) {
  private val owner: WeakReference<Entity>? = ownerEntity
    ?.also { customName = it.name }
    ?.let { WeakReference(it) }

  private var digPosition: BlockPos? = null
  private var digBlock: Block? = null

  private var currentDamage = -1
  private var currentDamageState = -1

  init {
    (this as EntityAccessor).setStandingEyeHeight(0.0f)
  }

  override fun getHoverEvent(): HoverEvent = owner?.get()
    ?.let { HoverEvent(SHOW_ENTITY, HoverEvent.EntityContent(it.type, it.uuid, it.name)) }
    ?: HoverEvent(SHOW_TEXT, Text.of("PlethoraFakePlayer - No owner!"))

  override fun getPitch(tickDelta: Float): Float = pitch // No pitch lerping
  override fun getYaw(tickDelta: Float): Float = yaw // No yaw lerping, and don't use head yaw like livingEntity does
  override fun getEyeY() = y
  override fun getEyeHeight(pose: EntityPose) = 0.0f
  override fun getActiveEyeHeight(pose: EntityPose, dimensions: EntityDimensions) = 0.0f

  fun updateCooldown() {
    handSwingTicks = 20
  }

  private fun setState(block: Block?, pos: BlockPos?) {
    val spim = interactionManager as ServerPlayerInteractionManagerAccessor
    spim.setMining(false)
    spim.blockBreakingProgress = -1

    digPosition = pos
    digBlock = block
    currentDamage = -1
    currentDamageState = -1
  }

  fun dig(pos: BlockPos, side: Direction?): Pair<Boolean, String> {
    val world = entityWorld
    val state = world.getBlockState(pos)
    val block = state.block
    val material = state.material

    val topY = world.topY

    if (block !== digBlock || pos != digPosition) setState(block, pos)

    if (!world.isAir(pos) && !material.isLiquid) {
      if (block === Blocks.BEDROCK || state.getHardness(world, pos) <= -1) {
        return Pair(false, "Unbreakable block detected")
      }

      val spim = interactionManager as ServerPlayerInteractionManagerAccessor
      for (i in 0..9) {
        if (currentDamageState == -1) {
          // TODO: block breaking is now sequenced. may need to increment the sequence each time
          interactionManager.processBlockBreakingAction(pos, START_DESTROY_BLOCK, side, topY, 0)
          currentDamageState = spim.blockBreakingProgress
        } else {
          currentDamage++

          val hardness = state.calcBlockBreakingDelta(this, world, pos) * (currentDamage + 1)
          val hardnessState = (hardness * 10).toInt()
          if (hardnessState != currentDamageState) {
            world.setBlockBreakingInfo(id, pos, hardnessState)
            currentDamageState = hardnessState
          }

          if (hardness >= 1) {
            interactionManager.tryBreakBlock(pos)
            setState(null, null)
            break
          }
        }
      }

      return Pair(true, "block")
    }

    return Pair(false, "Nothing to dig here")
  }

  companion object {
    val PROFILE = GameProfile(FAKEPLAYER_UUID, "[$modId]")
  }
}
